### PR TITLE
Implement MSC3973: Search users in the user directory with the Widget API

### PR DESCRIPTION
**Pull request by dhenneke**
https://github.com/matrix-org/matrix-spec-proposals/pull/3973

Depends on:

* https://github.com/matrix-org/matrix-widget-api/pull/79

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [ ] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

Type: enhancement

---
Here's what your changelog entry will look like:

## ✨ Features
 * Implement MSC3973: Search users in the user directory with the Widget API ([\#10269](https://github.com/matrix-org/matrix-react-sdk/pull/10269)). Contributed by @dhenneke.

### CODE REVIEW

File: src/stores/widgets/StopGapWidgetDriver.ts [hunk 1]
```diff
@@ -485,4 +486,19 @@ export class StopGapWidgetDriver extends WidgetDriver {
             prevBatch: prevBatch ?? undefined,
         };
     }
+
+    public async searchUserDirectory(searchTerm: string, limit?: number): Promise<ISearchUserDirectoryResult> {
+        const client = MatrixClientPeg.get();
+
+        const { limited, results } = await client.searchUserDirectory({ term: searchTerm, limit });
```
Comment (dhenneke) [inline src/stores/widgets/StopGapWidgetDriver.ts:493]: An alternative here could be to refactor and reuse the code in the [`InviteDialog`](https://github.com/matrix-org/matrix-react-sdk/blob/e5291c195d3259b5905d72dc2945ed740fffc6d1/src/components/views/dialogs/InviteDialog.tsx) to have the additional logic of making the search experience more personalised/useful. But I'm not sure if this is crucially important (see also https://github.com/matrix-org/matrix-spec-proposals/pull/3973#pullrequestreview-1320398823).

Comment (dhenneke) [reply src/stores/widgets/StopGapWidgetDriver.ts:494]: As discussed in the MSC, we pass though the raw search results from the server. So there is no need for a more complex logic.

**Comment by dhenneke** [comment]
I updated `matrix-widget-api` to version 1.3.0 and finished the implementation. After this PR, the `matrix-widget-api` should also be updated in the `matrix-js-sdk` as well as `element-web`.

**Review by andybalaam — APPROVED** [review]
Looks good, thanks!

**Comment by andybalaam** [comment]
@dhenneke please can you resolve the conflicts and I will merge? Thanks!

**Comment by dhenneke** [comment]
Thanks for the approval 🎉

I resolved the conflict with https://github.com/matrix-org/matrix-react-sdk/pull/10409 that already updated the dependency to a newer version.

**Comment by andybalaam** [comment]
> Thanks for the approval tada
> 
> I resolved the conflict with #10409 that already updated the dependency to a newer version.

Thank you!

**Pull request merged**